### PR TITLE
[SERVICE-808] Perform "App developer" testing of notifications client

### PR DESCRIPTION
**Pull request by bryangaleOF**
This PR
- Implements database migration so older DBs are properly handled (any pre-Dexie DBs are ignored entirely)
- Improves error messages on type errors
- Brings over ErrorHandler changes from FDC3
- Upgrades TS version (needed for `Omit` used by `PartiallyWritable`)
- Creates a `validation.ts` file, modeled after FDC3's
- Improves the example in the README

This PR does not
- Add integration tests for DB migration (this needs some kind of "start provider with this DB snapshot" functionality)
- Fully unify error messages with FDC3
- Prevent devs from importing types from `internal.d.ts` (will look into as part of FDC3 types story)

### CODE REVIEW

File: src/provider/model/database/Database.ts [hunk 1]
```diff
@@ -31,17 +33,83 @@ export class Database extends AsyncInit {
...
     }
 
     protected async init(): Promise<void> {
-        await this._database.open();
+        try {
+            await this._database.open();
+        } catch (e) {
+            // Version 1 database may or may not have the Applications store, so we need to handle both cases
+            if (e instanceof Dexie.DexieError) {
+                console.warn('Failed to open database, reattempting with alternative version 1 schema');
```
[the remark this answers is not in the document]
Comment (bryangaleOF) [reply src/provider/model/database/Database.ts]: > Assuming the code below will re-define version 1, but still keep version 2 and it's migration code?

Yep, that's correct and I've tested that scenario

File: src/provider/index.ts [hunk 2]
```diff
@@ -155,7 +156,11 @@ export class Main {
      * @returns Whether or not the removal of the notifications was successful.
      */
     private async clearNotification(payload: ClearPayload, sender: ProviderIdentity): Promise<boolean> {
-        const id = this.encodeID(payload.id, sender);
+        if (typeof payload.id !== 'string') {
+            throw new Error('Invalid argument passed to clear: argument must be a string');
```
[the remark this answers is not in the document]
Comment (bryangaleOF) [reply src/provider/index.ts]: We're generally doing provider side checking, unless there's a specific reason to do client side. Any reason here?

File: src/provider/index.ts [hunk 3]
```diff
@@ -249,10 +254,18 @@ export class Main {
             problems.push('notifications can have at-most four buttons');
         }
 
+        if (payload.buttons !== undefined) {
+            if (payload.buttons.some((button: ButtonOptions) => typeof button !== 'object' || button === null)) {
+                problems.push('"buttons" must be an array of objects');
```
[the remark this answers is not in the document]
Comment (bryangaleOF) [reply src/provider/index.ts]: ✅

File: package.json [hunk 4]
```diff
@@ -75,7 +75,7 @@
     "redux-devtools-cli": "0.0.1-1",
     "reflect-metadata": "^0.1.13",
     "semver-compare": "^1.0.0",
-    "typescript": "3.4.5"
+    "typescript": "3.7.5"
```
Comment (ryanpetersonOF) [inline package.json]: I'm happy this works without any major issues

File: src/provider/index.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (pjbroadbent) [reply src/provider/index.ts]: Guess I was thinking of FDC3 :)

**Pull request merged**